_drafts/autolab-and-docker.markdown

```diff
@@ -0,0 +1,56 @@
+
+While these events have been instrumental in making Autolab usable, the time and effort that needs to be spent in setting up and configuring Tango, the autograder, is non-trivial. Furthermore, Tango supports a limited number of platforms (Local, EC2) on which it can run jobs. Hence, as part of the latest efforts in making Autograding for all a reality, we are [integrating Docker][gh-docker] with Autolab in order to significantly reduce the overhead in getting started and to make autograding platform-agnostic. 
+
+This post starts by shedding some light on the concepts Tango uses for autograding. Building on those concepts, this post will then outline how docker is integrated with Tango.
+
+# What it takes to Autograde
+
+A guarantee that Tango must provide is that the environment in which a job runs must remain homogeneous between jobs. Hence, Tango executes these jobs on Virtual Machines that are booted with user-specified images. In order to run jobs on a particular platform such as Amazon EC2, OpenStack or even RackSpace, the VMM interface must be implemented for that platform. The [VMM interface][gh-sample-code] is as follows:
+
+* `initializeVM` - Initializes a VM for that platform.
+* `waitVM` - Wait for a VM to be accessible
+* `copyIn` - Copy all necessary input files to run the job.
+* `runJob` - Invoke the autodriver to run the job. The autodriver instruments the job, runs it in a controlled environment and restores the system state after job completion.
+* `copyOut` - Copy feedback file from the VM.
+* `destroyVM` - Destroy a given VM.
+* `getVMs` - Get a list of VMs in each VM pool.
+
+At present, the VMM interface has been implemented for any local Linux machine and Amazon EC2. In order to execute a job, Tango's job manager must check if there is an available VM to run the job (`getVMs`), ensure that the VM is reachable (`waitVM`), followed by `copyIn`, `runJob` and `copyOut`.
+
+# Integrating Docker
+
+Docker containers provide process, network and memory isolation. They are also designed to be multi-tenant citizens on any host they run on. A combination of these properties makes docker containers ideal for running autograding jobs. Furthermore, clients have the flexibility of [building any Docker image][docker-img] with customized software and running jobs in those images. Tango's [v1.1.0 release][gh-docker] allows clients to run jobs in docker containers on the same host as Tango, i.e. locally. This option is useful for small classes and needs minimal infrastructure.
+
+In the context of docker, here is what the VMM interface does under the hood:
+
+* `initializeVM` - Nothing to initialize!
```

Comment:
It might be worth mentioning that there's nothing to initialize because Docker containers are designed to be ephemeral as well as available on demand.